### ISSUE TITLE
Option to include related targets (such as tests) when focusing on a target

### ISSUE THREAD

**Issue by hiltonc** [tuist/tuist#6512]
### What problem or need do you have?

When focusing on a target (`tuist generate MyTarget`) I think it would make sense to also include related test targets (`MyTargetTests`). Currently I need to list them explicitly (`tuist generate MyTarget MyTargetTests`).

### Potential solution

We could add a `--include-name-suffixes <name-suffixes> (env: TUIST_GENERATE_INCLUDE_NAME_SUFFIXES)` argument to the `generate` command. This would cause the list of generation targets to be augmented by the target names with each of the suffixes.

For example:

```
tuist generate Common Log --include-name-suffixes Tests IntegrationTests
```

would focus on:

```
Common
CommonTests (if it exists)
CommonIntegrationTests (if it exists)
Log
LogTests (if it exists)
LogIntegrationTests (if it exists)
```

In practice, I would just add this to my development shell:

```
export TUIST_GENERATE_INCLUDE_NAME_SUFFIXES=Tests IntegrationTests
```

And then do `tuist generate Common Log` and get the same result.

### macOS version

14.5

### Tuist version

4.20.0

### Xcode version

15.4

**Comment by pepicrft** [comment]
Thanks @hiltonc. This has come up a few times. I think what we should do here is:
1. Include the tests because they are strongly connected to the target being focused on.
2. Allow associating metadata to targets that users can then use to group focusable targets as they wish:

```swift
let target = Target.target(name: "MyTarget", metadata: ["team": "foo"])
```

We can start with 1, and do 2 later. Would you like to take a stab at implementing this?

**Comment by hiltonc** [comment]
@pepicrft I'd be happy to attempt this. Do you want to use the `--include-name-suffixes Tests` approach, or should tests always be automatically included?

**Comment by pepicrft** [comment]
Tests targets should be included automatically. Given a target, you can get upstream test targets that depend on those, and include them.
After doing that, if you want to push this a bit further and provide some flexibility, we can then add support for something like:

```
tuist generate tag:foo tag:bar
```

**Comment by hiltonc** [comment]
@pepicrft I tried doing this but ran into a wall. I can include a `var metadata: [String: String]` in `ProjectDescription.Target`, but by the time we're in `GraphTraversing.filterIncludedTargets` we're working with `XcodeGraph.Target`. It wouldn't be appropriate to add `metadata` to `XcodeGraph`, would it? Maybe I could populate a map in the `MapperEnvironment`?

**Comment by pepicrft** [comment]
> @pepicrft I tried doing this but ran into a wall. I can include a var metadata: [String: String] in ProjectDescription.Target, but by the time we're in GraphTraversing.filterIncludedTargets we're working with XcodeGraph.Target. It wouldn't be appropriate to add metadata to XcodeGraph, would it? Maybe I could populate a map in the MapperEnvironment?

That's correct. We'd need to add it to the interface, and also to the equivalent model in `XcodeGraph`. Then we can use that information for filtering.
